README.md
# cURL Request for Node.js

[![License][npm-license]][license-url]
[![Downloads][npm-downloads]][npm-url]

cURL Request is a cURL wrapper, support all options and infos from cURL.

## Why a cURL wrapper?

* All node libaries i have tested are very buggy in case you use a proxy and https
* cURL is working for all cases since years
* The wrapper is easy to use with promises 

## Installing

To utilize for node.js install the the `npm` module:

```bash
$ npm install curl-request --save
```

After installing the `npm` package you can now start simplifying requests like so:

```js
const curl = new (require( 'curl-request' ))();
```

If you have problems installing the dependencies, use this to build it from source:

```bash
$ npm install node-libcurl --build-from-source 
```

## Usage

```js
const curl = new (require( 'curl-request' ))();

curl.setHeaders([
    'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36'
])
.get('https://www.google.com')
.then(({statusCode, body, headers}) => {
    console.log(statusCode, body, headers)
})
.catch((e) => {
    console.log(e);
});
```

```js
curl
.setBody({
 'input-arr[0]': 'input-arr-val0',
 'input-arr[1]': 'input-arr-val1',
 'input-arr[2]': 'input-arr-val2',
 'input-name': 'input-val'
})
.post('https://www.google.com')
.then(({statusCode, body, headers}) => {
    console.log(statusCode, body, headers)
})
.catch((e) => {
    console.log(e);
});
```

```js
// File upload
curl
.setHeaders([
    'Content-Type: multipart/form-data'
])
.setMultipartBody([{
  name: 'filename',
  contents: 'yourimage.png'
}, {
  name: 'file',
  file: './yourimage.png',
  type: 'image/png'
}])
.post('https://www.google.com')
.then(({statusCode, body, headers}) => {
    console.log(statusCode, body, headers)
})
.catch((e) => {
    console.log(e);
});
```

```js
// Set tor proxy
curl.default.useProxy = true;
curl.default.proxy =  'localhost:9050';
curl.default.proxyType = curl.libcurl.proxy.SOCKS5_HOSTNAME;
curl.default.torControlHost = 'localhost';
curl.default.torControlPort = 9051;
```

```js
// New tor identity
curl.newTorIdentity();
```

```js
// Verbose output
curl.default.verbose = true;
```

##
## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


----

[license-url]: https://github.com/do-web/curl-request/blob/master/LICENSE

[npm-url]: https://www.npmjs.com/package/curl-request
[npm-license]: https://img.shields.io/npm/l/curl-request.svg?style=flat
[npm-version]: https://badge.fury.io/js/curl-request.svg
[npm-downloads]: https://img.shields.io/npm/dm/curl-request.svg?style=flat

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## 1.1.1 - 2018-01-30

## 1.1.0 - 2018-01-30

### Added
- Changelog

### Changed
- module.exports now returns only a function. 
Now you must create a new instance, it prevents from overwriting the default options.

index.js
const Curl = require('node-libcurl').Curl;
const Net = require('net');
const querystring = require('querystring');

module.exports = function () {

    this['default'] = {
        torControlHost: 'localhost',
        torControlPort: 9051,
        autoParse: true, // content-type detect -> json
        verbose: false,
        useProxy: false,
        proxy: 'localhost:9050',
        proxyType: Curl.proxy.SOCKS5_HOSTNAME
    };

    this.libcurl = Curl;

    this.curl = new Curl();

    this.newTorIdentity = () => {
        let client = new Net.Socket();
        client.connect(this['default'].torControlPort, this['default'].torControlHost, () => {
            client.write('authenticate\nsignal newnym\nquit');
        });
    };

    this._setUrl = (url) => {
        this.curl.setOpt(Curl.option.URL, url);
        return this;
    };

    this.setMultipartBody = (fieldsArray) => {
        this.curl.setOpt(Curl.option.HTTPPOST, fieldsArray);
        return this;
    };

    this.setOpt = (opt, val) => {
        this.curl.setOpt(opt, val);
        return this;
    };

    this.getCurl = () => {
        return this.curl;
    };

    this.setCurl = (curl_replace) => {
        this.curl = curl_replace;
        return this;
    };

    this.setBody = (fieldsObj) => {

        if (typeof fieldsObj !== 'string') {
            fieldsObj = querystring.stringify(fieldsObj)
        }
        this.curl.setOpt(Curl.option.POSTFIELDS, fieldsObj);
        return this;
    };

    this.setProxy = (host, proxyType) => {
        this.curl.setOpt(Curl.option.PROXY, host);

        // SOCKS5 default
        if (typeof proxyType === 'undefined') {
            this.curl.setOpt(Curl.option.PROXYTYPE, Curl.proxy.SOCKS5_HOSTNAME);
        } else {
            this.curl.setOpt(Curl.option.PROXYTYPE, proxyType);
        }

        return this;
    };

    this.setFollowLocation = (followlocation) => {
        this.curl.setOpt(Curl.option.FOLLOWLOCATION, followlocation);
        return this;
    };

    this.setHeaders = (headers) => {
        this.curl.setOpt(Curl.option.HTTPHEADER, headers);
        return this;
    };

    this.get = (url) => {
        this._setUrl(url);
        this.curl.setOpt(Curl.option.CUSTOMREQUEST, 'GET');
        return this._submit();
    };

    this.post = (url) => {
        this._setUrl(url);
        this.curl.setOpt(Curl.option.CUSTOMREQUEST, 'POST');
        return this._submit();
    };

    this.patch = (url) => {
        this._setUrl(url);
        this.curl.setOpt(Curl.option.CUSTOMREQUEST, 'PATCH');
        return this._submit();
    };

    this['delete'] = (url) => {
        this._setUrl(url);
        this.curl.setOpt(Curl.option.CUSTOMREQUEST, 'DELETE');
        return this._submit();
    };

    this.head = (url) => {
        this._setUrl(url);
        this.curl.setOpt(Curl.option.CUSTOMREQUEST, 'HEAD');
        return this._submit();
    };

    let normalizeHeaders = (headers) => {
        // normalize headers
        let nHeaders = {};
        if (Array.isArray(headers)) {
            let mergedHeaders = {};
            for(let i in headers) {
                mergedHeaders = Object.assign(mergedHeaders, headers[i]);
            }
            for(let k in mergedHeaders) {
                nHeaders[k.toString().toLocaleLowerCase()] = mergedHeaders[k];
            }
        }
        return nHeaders;
    };

    this._reset = () => {
        this.curl = new Curl();
    };

    this._submit = () => {

        this.curl.setOpt(Curl.option.VERBOSE, this['default'].verbose);

        if (this['default'].useProxy) {
            this.setProxy(this['default'].proxy, this['default'].proxyType);
        }

        return new Promise((resolve, reject) => {

            try {
                this.curl.on('end', (statusCode, body, headers) => {

                    headers = normalizeHeaders(headers);

                    if (this.default.autoParse) {
                        if (typeof headers['content-type'] !== 'undefined' &&
                            headers['content-type'].toLocaleLowerCase() === 'application/json') {
                            try {
                                let jsonObj = JSON.parse(body);
                                body = jsonObj;
                            } catch (e) {
                            }
                        }
                    }

                    this.curl.close();
                    this._reset();
                    resolve({statusCode, body, headers});
                });

                this.curl.on('error', () => {
                    this.curl.close();
                    this._reset();
                    reject(arguments);
                });
                this.curl.perform();
            } catch (e) {
                reject(e);
            }
        });
    };

    return this;
};